File: src/api/settings/ui.ts
import { RawDictionary } from "../../i18n/i18n.ts";
import { SettingsT } from "../settings.ts";

export interface Settings {
  sections: Section[];
}

export interface Section {
  id: keyof RawDictionary["settings"]["section"];
  settings: Setting[];
}

export interface ToggleSetting {
  key: SettingsT<boolean>;
  input: "toggle";
}

export interface TextSetting {
  key: SettingsT<string>;
  input: "text";
}

export interface GameSelectSetting {
  key: SettingsT<string>;
  input: "game_select";
}

export type Setting = ToggleSetting | TextSetting | GameSelectSetting;

File: src-tauri/src/settings/ui.rs
#[derive(serde::Serialize)]
pub struct Settings {
    pub sections: &'static [Section],
}

#[derive(serde::Serialize)]
pub struct Section {
    pub id: &'static str,
    pub settings: &'static [Setting],
}

#[derive(serde::Serialize)]
pub struct Setting {
    pub key: &'static str,
    pub input: Input,
}

#[derive(serde::Serialize)]
#[serde(tag = "type")]
pub enum Input {
    Toggle,
    Text,
}
